lib/utils/prompt/context_reader.go

```diff
@@ -184,6 +185,41 @@ func (cr *ContextReader) processReads() {
 	}
 }
 
+// handleInterrupt restores terminal state on interrupts.
+// Called only on global ContextReaders, such as Stdin.
+func (cr *ContextReader) handleInterrupt() {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	defer signal.Stop(c)
+
+	for {
+		select {
+		case sig := <-c:
+			log.Debugf("Captured signal %s, attempting to restore terminal state", sig)
+			cr.mu.Lock()
+			_ = cr.maybeRestoreTerm(iAmHoldingTheLock{})
+			cr.mu.Unlock()
+		case <-cr.closed:
+			return
+		}
+	}
+}
+
+// iAmHoldingTheLock exists only to draw attention to the need to hold the lock.
+type iAmHoldingTheLock struct{}
```

Comment:
Haven't seen this one before. In most places we just suffix the method name with `Locked` to indicate that you should hold the lock before calling.

Not that there's anything wrong with this. It's just different.